### PR TITLE
8273681: Add Vector API vs Arrays.mismatch intrinsic benchmark

### PR DESCRIPTION
**Pull request by amCap1712**
Hi all!

Please review this PR to add a benchmark comparing the performance of Arrays.mismatch intrinsic in the JDK with that of the Vector API. Kindly refer to this [thread] on panama-dev regarding some initial discussion about this benchmark. I have attached the [results] of the full benchmark run along with the [assembly] output of a shorter run I had done while analysing the results. The benchmarks were run against the latest build of panama-vector available from builds.shipilev.net.

Also, I have not added the copyright header to this file yet as I am an individual contributor (OCA signed) and do not know what to put there.

[thread]: https://mail.openjdk.java.net/pipermail/panama-dev/2021-September/014839.html
[results]: https://github.com/openjdk/jdk/files/7142452/results.csv
[assembly]:  https://github.com/openjdk/jdk/files/7142362/benchmarks.txt

Regards,
Kartik

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273681](https://bugs.openjdk.java.net/browse/JDK-8273681): Add Vector API vs Arrays.mismatch intrinsic benchmark

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**) ⚠️ Review applies to b05778070623d6e9bea3f6d3d9dd7dadd286fcb8

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5459/head:pull/5459` \
`$ git checkout pull/5459`

Update a local copy of the PR: \
`$ git checkout pull/5459` \
`$ git pull https://git.openjdk.java.net/jdk pull/5459/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5459`

View PR using the GUI difftool: \
`$ git pr show -t 5459`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5459.diff">https://git.openjdk.java.net/jdk/pull/5459.diff</a>

</details>

### CODE REVIEW

**Comment by PaulSandoz** [comment]
The simplest approach is to copy the license from here https://github.com/openjdk/jdk/blob/master/test/micro/org/openjdk/bench/jdk/incubator/vector/RotateBenchmark.java

File: test/micro/org/openjdk/bench/jdk/incubator/vector/ArrayMismatchBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+    float[] floatData1;
+    float[] floatData2;
+
+    static final VectorSpecies<Byte> BYTE_SPECIES_PREFERRED = ByteVector.SPECIES_PREFERRED;
+    static final VectorSpecies<Integer> INT_SPECIES_PREFERRED = IntVector.SPECIES_PREFERRED;
+    static final VectorSpecies<Float> FLOAT_SPECIES_PREFERRED = FloatVector.SPECIES_PREFERRED;
+    static final VectorSpecies<Long> LONG_SPECIES_PREFERRED = LongVector.SPECIES_PREFERRED;
+
+    static final Random random = new Random();
```
Comment (PaulSandoz) [inline test/micro/org/openjdk/bench/jdk/incubator/vector/ArrayMismatchBenchmark.java:52]: We could use the recently added `RandomGenerator` instead in the spirit of encouraging the use of new and preferred APIs:

- remove the static field
- replace `FLOAT_SPECIES` with `DOUBLE_SPECIES`
- in `setup` create an instance `RandomGenerator rg = RandomGenerator.getDefault()`
- remove `createRandomFloats` and use `rg.doubles(...).toArray()`

**Review by PaulSandoz — COMMENTED** [review]
Benchmark looks good (assuming license is added).

File: test/micro/org/openjdk/bench/jdk/incubator/vector/ArrayMismatchBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            intData2 = random.ints(size).toArray();
+
+            int[] commonInts = random.ints(common).toArray();
+            System.arraycopy(commonInts, 0, intData1, 0, common);
+            System.arraycopy(commonInts, 0, intData2, 0, common);
+        } else if (params.getBenchmark().endsWith("Double")) {
+            doubleData1 = random.doubles(size).toArray();
+            doubleData2 = random.doubles(size).toArray();
+
+            double[] commonDoubles = random.doubles(size).toArray();
```
Comment (PaulSandoz) [inline test/micro/org/openjdk/bench/jdk/incubator/vector/ArrayMismatchBenchmark.java:103]:
```suggestion
            double[] commonDoubles = random.doubles(common).toArray();
```

**Review by PaulSandoz — APPROVED** [review]
Looks good, just a minor issue (that does not induce incorrect behavior), no need for re-approval to fix that.

File: test/micro/org/openjdk/bench/jdk/incubator/vector/ArrayMismatchBenchmark.java (hunk 2 above)
Comment (amCap1712) [reply test/micro/org/openjdk/bench/jdk/incubator/vector/ArrayMismatchBenchmark.java]: Fixed. Thanks!

**Comment by amCap1712** [comment]
Thanks for the review, @PaulSandoz
/integrate

**Pull request closed without merging**